pybind11_configure.bzl

```diff
@@ -0,0 +1,27 @@
+"""Module extension for "configuring" pybind11."""
+
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+def _parse_my_own_version_from_module_dot_bazel(ctx):
+    lines = ctx.read(Label("//:MODULE.bazel")).split("\n")
+    for line in lines:
+        parts = line.split("\"")
+        if parts[0] == "    version = ":
+            return parts[1]
+    fail("Failed to parse my own version from `MODULE.bazel`! " +
+         "This should never happen!")
+
+def _pybind11_configure_extension_impl(ctx):
+    version = _parse_my_own_version_from_module_dot_bazel(ctx)
+
+    # The pybind11_bazel version should typically just be the pybind11 version,
+    # but can end with ".bzl.<N>" if the Bazel plumbing was updated separately.
+    version = version.split(".bzl.")[0]
+    http_archive(
+        name = "pybind11",
+        build_file = "//:pybind11.BUILD",
+        strip_prefix = "pybind11-%s" % version,
+        urls = ["https://github.com/pybind/pybind11/archive/v%s.zip" % version],
+    )
```

Comment:
What is all this about...?

It looks like pybind11_bazel is already in the bazel registry, so you shouldn't need to do this...er, but where is the config that publishes to the registry? How did it get into the registry? Where they manually crafted and put in...?

I'm fairly certain this entire file can be done away with, insofar as bzlmod builds are concerned.

Reply:
pybind11_bazel is in BCR; pybind11 isn't. Thus, we fetch the corresponding version of pybind11 automagically. We have been doing this since https://github.com/pybind/pybind11_bazel/commit/af7151f632550cd2f470fdefacf17770d5b94959. If you have a better idea, I'm open to refining this! (Ideally in a followup PR?)

Reply:
Ahhhh, I missed that part. Yeah, that's annoying.

Ok, then what you want to do is move this into pybind11_bazel's MODULE file. This lets pybind11_bazel manage dependencies directly and users don't have to do anything in their module files.

I recommend the following:

```
# MODULE.bazel
version = "X"
module(name="pybind11_bazel", version=version)

setup = use_extension("//:internal_setup.bzl", "setup")
setup.pybind11(pybind11_version = version)

# internal_setup.bzl

setup = module_extension(
  implementation = _setup_impl,
  tag_classes = {"pybind11": tag_class(attrs = {
    "pybind11_version": attr.string(),
  }}
)

def setup_impl(mctx):
  pybind11_version = mctx.tags.setup[0].pybind11_version
  http_archive(...use version...)
```

You can decouple the pybind11_version from pybind11 version if you want (just don't re-use the `version` variable). This can also be easily adapted to let users customize the pybind11 archive/version they use (just make the `setup.pybind11()` extension/tag public).

Reply:
Thanks! But will that obviate `use_repo(pybind11_configure, "pybind11")` in users' `MODULE.bazel` files though? 🤔

Reply:
Oh, I guess we could move everything from `pybind11.BUILD` to `BUILD` and provide a `filegroup()` in the former to proxy the headers to the latter? Although I would **_really_** prefer to do all of this in a followup PR. ;)

Reply:
Keep the `http_archive(...)` part as-is (e.g. continue to use `build_file = "//pybind.BUILD"`).

Yeah, you _could_ remove pybind.BUILD, but it doesn't really gain you much. The point of http_archive is to take a bunch of files from some external place and make them look like a regular repository, which is exactly what you're doing and want.

The point of my snippet is it lets pybind11_bazel do all the setup itself; users don't need to do anything other than `bazel_dep()`.

> will that obviate use_repo(pybind11_configure, "pybind11") in users' MODULE.bazel files

Yes. A user would only need to do that if they themselves need to directly reference `@pybind//<something>`. Which they probably shouldn't be doing; your pybind.BUILD file is just for pybind11_bazel's usage, right? It's not really intended to be part of a public api to make arbitrary parts of pybind11 itself available to users? And, like I said on chat, if they need to, you can't really stop them from doing so, and it'll probably aid workspace->bzlmod transition.

In chat I talked about using toolchains and letting users customize the version -- Yeah, I think its fine to not worry about that in this PR. Let me know on chat, and I can sketch out how to do this. It's pretty straight forward and mostly boilerplate type of code to wire up bzlmod stuff.

Reply:
Thanks! https://github.com/pybind/pybind11_bazel/issues/71#issuecomment-1939537544 captures my intention to do this. :)